lib/firenest/topology/erlang.ex

```diff
@@ -261,53 +277,60 @@ defmodule Firenest.Topology.Erlang.Server do
 
   # If two nodes come up at the same time, ping may be received twice.
   # So we need to make sure to handle two pings/pongs.
-  def handle_info({:ping, other_id, pid, monitors}, %{id: id} = state) do
-    pong(pid, id, Map.to_list(state.monitors))
-    {:noreply, add_node(state, other_id, pid, monitors)}
+  def handle_info({:ping, pid, other_id, clock, monitors}, state) do
+    pong(state, pid)
+    {:noreply, add_node(state, pid, other_id, clock, monitors)}
   end
 
-  def handle_info({:pong, other_id, pid, monitors}, state) do
-    {:noreply, add_node(state, other_id, pid, monitors)}
+  def handle_info({:pong, pid, other_id, clock, monitors}, state) do
+    {:noreply, add_node(state, pid, other_id, clock, monitors)}
   end
 
   ## Helpers
 
-  defp ping(node, topology, id, monitors) when is_list(monitors) do
-    Process.send({topology, node}, {:ping, id, self(), monitors}, [:noconnect])
+  defp ping(state, node) do
+    %{topology: topology, id: id, clock: clock, monitors: monitors} = state
+    monitors = Map.to_list(monitors)
+    Process.send({topology, node}, {:ping, self(), id, clock, monitors}, [:noconnect])
   end
 
-  defp pong(pid, id, monitors) when is_list(monitors) do
-    Process.send(pid, {:pong, id, self(), monitors}, [:noconnect])
+  defp pong(state, pid) do
+    %{id: id, clock: clock, monitors: monitors} = state
+    monitors = Map.to_list(monitors)
+    Process.send(pid, {:pong, self(), id, clock, monitors}, [:noconnect])
   end
 
   defp id() do
     {:crypto.strong_rand_bytes(4), System.system_time()}
   end
 
-  defp add_node(%{nodes: nodes} = state, id, pid, monitors) do
+  defp add_node(%{nodes: nodes} = state, pid, id, clock, monitors) do
     node = Kernel.node(pid)
     new_remote_names = for {ref, name} <- monitors, do: {name, ref}, into: %{}
 
     case nodes do
-      %{^node => {^id, ref, old_remote_names}} ->
+      %{^node => {^id, ^clock, _, _}} ->
+        state
+
+      %{^node => {^id, _, ref, old_remote_names}} ->
```

Comment:
I think we should remove this branch as to get here we got out of sync/node down issue, we can simplify and fallback to next branch.